File: leetcode/python/Insertion Sort List.py
# Insertion Sort List
# for leetcode problems
# 2014.08.15 by zhanglin

# Problem:
# Sort a linked list using insertion sort.

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def insertionSortList(self, head):
        if head == None:
            return head

        dummy = ListNode(0)
        dummy.next = head

        current = head
        while current.next:
            if current.val < current.next.val: # if sorted, move forward
                current = current.next

            else: # this node need to be inserted to previous sorted list
                insertPos = dummy # find insertion position from the very beginning
                while insertPos.next.val < current.next.val: # find appropriate position
                    insertPos = insertPos.next

                # after while loop, now insertPos.val <= TobeMoved.val <= insertPos.next.val
                # then insert TobeMoved between insertPos and insertPos.next
                TobeMoved = current.next
                current.next = TobeMoved.next
                TobeMoved.next = insertPos.next
                insertPos.next = TobeMoved

        return dummy.next

File: leetcode/c++/Insertion Sort List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode *insertionSortList(ListNode *head)
    {
        ListNode *new_head = new ListNode(0);
        ListNode *pt_cache = NULL;
        ListNode *pt_currt = NULL;

        if(head == NULL || head->next == NULL)
        {
            return head;
        }

        while(head != NULL)
        {
            if(new_head->next == NULL)
            {
                new_head->next = head;
                head = head->next;
                new_head->next->next = NULL;
            }
            else
            {
                pt_cache = new_head;
                pt_currt = new_head->next;

                while(pt_currt != NULL)
                {
                    if(head->val <= pt_currt->val)
                    {
                        pt_cache->next = head;
                        head = head->next;
                        pt_cache->next->next = pt_currt;
                        break;
                    }
                    else
                    {
                        pt_cache = pt_cache->next;
                        pt_currt = pt_currt->next;
                        continue;
                    }
                }

                if(pt_currt == NULL)
                {
                    pt_cache->next = head;
                    head = head->next;
                    pt_cache->next->next = NULL;
                }
            }
        }
        return new_head->next;
    }
};